### ISSUE TITLE
Doesn't work properly with Loupedeck CT

### ISSUE THREAD

**Issue by scottlaird** [scottlaird/loupedeck#1]
Loupedeck currently makes 4 models:

- Loupedeck Live S
- Loupedeck Live
- Loupedeck+
- Loupedeck CT

This code was developed for the Live and doesn't currently support any of the others, but it *probably* isn't that hard to get them working.

I have a Loupedeck CT in addition to the Loupedeck Live, so I decided to give it a spin and see what happens.  It helps that the CT is basically a superset of the Live, so my app should reasonably continue to work with it.  The Loupedeck+ doesn't have a display, and the Loupedeck Live S lacks several of the buttons that I depend on, so they'd be harder to support.

The CT had a few issues: 

1.  It took multiple attempts before it connected correctly with WebSockets, but it did connect eventually.
2. Nothing appeared on the display.  The text that should have been visible was not.
3. Controls actually worked.  I was working blind, but all of the basic functionality worked.  The 6 small dials worked, the touchscreen worked, and the row of buttons under the touchscreen worked.
4. Additional buttons sent usable button codes, which were logged as unknown (as expected)
5. Nothing was logged when the big central dial was turned.
6. Events were logged when the central dial's touchscreen was touched.

The first problem to tackle will be #2, nothing being shown on the display.  After that, I'll update the code to add constants for the additional buttons, and see if I can figure out the websocket problem.  Finally, I'll look into the big dial.

**Comment by scottlaird** [comment]
Hmm.  Looking at https://github.com/foxxyz/loupedeck, it looks like this *should* work.  The CT is almost a superset of the Live.  In [their code](https://github.com/foxxyz/loupedeck/blob/5d092a4e246b0a7ae1558678c655f12d82b215c6/device.js#L295), the `LoupedeckCT` class is just an extension of the `LoupedeckLive` class with a different product ID, a few more buttons, an extra display, and an extension to the touch display coordinate decoding logic.

The 3 main screens are the same size, same resolution, and go by the same names in the protocol.  So the protocol messages that update the screens should be the same across both devices.  As likely as not, I just have a stupid bug somewhere that the firmware rev on my Live accepts and the firmware rev on the CT doesn't.

**Comment by scottlaird** [comment]
It looks like there are 2 different versions of the Loupedeck CT, the original (product ID 0003) and "v2" (product ID 0007).  According to https://github.com/foxxyz/loupedeck/issues/27, the v2 has a different screen layout in software (1 big screen instead of 3 side-by-side screens).  That would explain my issues, but making trivial changes to draw to screen ID 'M' instead of screen ID 'A' still don't result in anything being drawn on the screen.

**Comment by scottlaird** [comment]
Making a bit of progress, but still nothing on the CT's screen.

With the Live, when I draw onto the screen with the `WriteFramebuff` (`0x10`) and `Draw` (`0x0f`) calls, I get a response back from the Loupedeck with a matching transaction ID and a body of `1`.  I'm going to take this as a "everything is fine" message.

With the CT, running the Live's code, both `WriteFramebuff` and `Draw` return `0`.  I'll guess that this is an error.

When I change the code to send the CT draw requests for display ID `M`, then I get a `WriteFramebuff` response of 1 (success?) and `Draw` of `0` (failure?).  Hmm.

Also, drawing to screen `W` (the one inside the big knob on the CT) crashes the CT if I try to write too far out of bounds.  So I can't just use it as a quick debugging tool by telling my code that the (say) center display's ID is `W`.

Here's what I see when I try to draw a 90-pixel square box on the main display at (0,0).  That translates to 60,0 on display `M` with the CT:

```
2023/10/08 13:32:25 INFO Draw called Display=main xoff=0 yoff=0 width=90 height=90
2023/10/08 13:32:25 INFO Draw parameters x=60 y=0 width=90 height=90
2023/10/08 13:32:25 INFO Sending message="{len: 255, type: 10, txn: 1c, data: [0 77 0 60 0 0 0 90 0 90 248 254 248 254 248 254]..., actual_len: 16210}"
2023/10/08 13:32:25 INFO Sending message="{len: 5, type: 0f, txn: 1d, data: [0 77]}"
2023/10/08 13:32:25 INFO Read message="{len: 4, type: 10, txn: 1c, data: [1]}"
2023/10/08 13:32:25 INFO Read message="{len: 4, type: 0f, txn: 1d, data: [0]}"
```

That translates into a `WriteFramebuff` call with transaction ID `1c`.  Data starts with `0 77`, which is `M`, then `0 60` (start at X=60), `0 0` (y=0), `0 90` (width=90), `0 90` (height=90), then a bunch of image bytes.  90*90*2 bytes = 16200 bytes, plus 10 bytes for the header gives 16210 bytes of data, which is what we're sending here in total.

It returns a `[1]` in response to the `WriteFramebuff` command (good?) and a `[0]` for the `Draw` command (bad?).  But the `Draw` command is trivial. 

I guess it's time to whip up a better test program.  I can write 1x1 pixel test images and then grow them slowly and see if it ever works.  I wouldn't be shocked to find a problem with the way I'm using websockets here.

**Comment by scottlaird** [comment]
Oh FFS.  Apparently the brightness defaults to 0 on the CTv2 or something?  Adding a call to SetBrightness seems to have fixed everything.  I used to have a call to SetBrightness in my code, but it was removed, apparently.

Going through CommandPost's [reset process](https://github.com/CommandPost/CommandPost/blob/974afaafb12e6d3572204a163a589df273989afb/src/extensions/hs/loupedeck/init.lua#L1399) (which supposedly works with the CTv2) showed that all that they do is set the brightness.

I'm still seeing `0` responses to `Draw` commands, but everything seems to be working, so maybe they don't matter?
